File: android/src/main/java/io/qonversion/capacitor/QonversionPlugin.kt
package io.qonversion.capacitor

import android.app.Activity
import android.app.Application
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import io.qonversion.sandwich.ActivityProvider
import io.qonversion.sandwich.QonversionEventsListener
import io.qonversion.sandwich.QonversionSandwich
import io.qonversion.sandwich.BridgeData

private const val EntitlementsUpdatedEvent = "entitlementsUpdatedEvent"

@CapacitorPlugin(name = "Qonversion")
class QonversionPlugin : Plugin() {
    private val qonversionSandwich by lazy {
        (context?.applicationContext as? Application)?.let {
            QonversionSandwich(
                it,
                object : ActivityProvider {
                    override val currentActivity: Activity?
                        get() = activity
                },
                qonversionEventsListener
            )
        } ?: throw IllegalStateException("Failed to initialize Qonversion Sandwich. Application is null.")
    }

    private val qonversionEventsListener: QonversionEventsListener = object :
        QonversionEventsListener {
        override fun onEntitlementsUpdated(entitlements: BridgeData) {
            notifyListeners(EntitlementsUpdatedEvent, entitlements.toJSObject())
        }
    }

    @PluginMethod
    fun initialize(call: PluginCall) {
        val context = context?.applicationContext ?: return call.reject("Can't get application context for Qonversion initialization", "InitializationError")
        val projectKey = call.getString("projectKey") ?: return call.noNecessaryDataError("projectKey")
        val launchModeKey = call.getString("launchMode") ?: return call.noNecessaryDataError("launchMode")
        val environmentKey = call.getString("environment") ?: return call.noNecessaryDataError("environment")
        val entitlementsCacheLifetimeKey = call.getString("entitlementsCacheLifetime") ?: return call.noNecessaryDataError("entitlementsCacheLifetime")
        val proxyUrl = call.getString("proxyUrl")
        val kidsMode = call.getBoolean("kidsMode") ?: false
        qonversionSandwich.initialize(context, projectKey, launchModeKey, environmentKey, entitlementsCacheLifetimeKey, proxyUrl, kidsMode)
        call.resolve()
    }

    @PluginMethod
    fun identify(call: PluginCall) {
        val userId = call.getString("userId") ?: return call.noNecessaryDataError("userId")

        qonversionSandwich.identify(userId, call.toResultListener())
    }

    @PluginMethod
    fun logout(call: PluginCall) {
        qonversionSandwich.logout()
        call.resolve()
    }

    @PluginMethod
    fun purchase(call: PluginCall) {
        val productId = call.getString("productId") ?: return call.noNecessaryDataError("productId")
        val oldProductId = call.getString("oldProductId")
        val offerId = call.getString("offerId")
        val applyOffer = call.getBoolean("applyOffer")
        val updatePolicyKey = call.getString("updatePolicyKey")
        val contextKeys = call.getArray("contextKeys")?.toList<String>()

        qonversionSandwich.purchase(
            productId,
            offerId,
            applyOffer,
            oldProductId,
            updatePolicyKey,
            contextKeys,
            call.toResultListener()
        )
    }

    @PluginMethod
    fun updatePurchase(call: PluginCall) {
        purchase(call)
    }

    @PluginMethod
    fun checkEntitlements(call: PluginCall) {
        qonversionSandwich.checkEntitlements(call.toResultListener())
    }

    @PluginMethod
    fun restore(call: PluginCall) {
        qonversionSandwich.restore(call.toResultListener())
    }

    @PluginMethod
    fun offerings(call: PluginCall) {
        qonversionSandwich.offerings(call.toResultListener())
    }

    @PluginMethod
    fun userProperties(call: PluginCall) {
        qonversionSandwich.userProperties(call.toResultListener())
    }

    @PluginMethod
    fun isFallbackFileAccessible(call: PluginCall) {
        qonversionSandwich.isFallbackFileAccessible(call.toResultListener())
    }

    @PluginMethod
    fun remoteConfig(call: PluginCall) {
        val contextKey = call.getString("contextKey")
        qonversionSandwich.remoteConfig(contextKey, call.toResultListener())
    }

    @PluginMethod
    fun remoteConfigList(call: PluginCall) {
        val contextKeys = call.getArray("contextKeys").toList<String>()

        if (contextKeys == null) {
            qonversionSandwich.remoteConfigList(call.toResultListener())
        } else {
            val includeEmptyContextKey = call.getBoolean("includeEmptyContextKey") ?: return call.noNecessaryDataError("includeEmptyContextKey")

            qonversionSandwich.remoteConfigList(contextKeys, includeEmptyContextKey, call.toResultListener())
        }
    }

    @PluginMethod
    fun syncHistoricalData(call: PluginCall) {
        qonversionSandwich.syncHistoricalData()
        call.resolve()
    }

    @PluginMethod
    fun products(call: PluginCall) {
        qonversionSandwich.products(call.toResultListener())
    }

    @PluginMethod
    fun setDefinedUserProperty(call: PluginCall) {
        val rawProperty = call.getString("property") ?: return call.noNecessaryDataError("property")
        val value = call.getString("value") ?: return call.noNecessaryDataError("value")

        qonversionSandwich.setDefinedProperty(rawProperty, value)
        call.resolve()
    }

    @PluginMethod
    fun setCustomUserProperty(call: PluginCall) {
        val property = call.getString("property") ?: return call.noNecessaryDataError("property")
        val value = call.getString("value") ?: return call.noNecessaryDataError("value")

        qonversionSandwich.setCustomProperty(property, value)
        call.resolve()
    }

    @PluginMethod
    fun syncPurchases(call: PluginCall) {
        qonversionSandwich.syncPurchases()
        call.resolve()
    }

    @PluginMethod
    fun addAttributionData(call: PluginCall) {
        @Suppress("UNCHECKED_CAST")
        val data = call.getObject("data") as? Map<String, Any> ?: return call.noNecessaryDataError("data")

        if (data.isEmpty()) {
            return call.noNecessaryDataError("data")
        }

        val provider = call.getString("provider") ?: return call.noNecessaryDataError("provider")

        qonversionSandwich.addAttributionData(provider, data)
        call.resolve()
    }

    @PluginMethod
    fun checkTrialIntroEligibility(call: PluginCall) {
        val ids = call.getArray("ids")?.toList<String>() ?: return call.noNecessaryDataError("ids")

        qonversionSandwich.checkTrialIntroEligibility(ids, call.toResultListener())
    }

    @PluginMethod
    fun attachUserToExperiment(call: PluginCall) {
        val experimentId = call.getString("experimentId") ?: return call.noNecessaryDataError("experimentId")
        val groupId = call.getString("groupId") ?: return call.noNecessaryDataError("groupId")

        qonversionSandwich.attachUserToExperiment(experimentId, groupId, call.toResultListener())
    }

    @PluginMethod
    fun detachUserFromExperiment(call: PluginCall) {
        val experimentId = call.getString("experimentId") ?: return call.noNecessaryDataError("experimentId")

        qonversionSandwich.detachUserFromExperiment(experimentId, call.toResultListener())
    }

    @PluginMethod
    fun attachUserToRemoteConfiguration(call: PluginCall) {
        val remoteConfigurationId = call.getString("remoteConfigurationId") ?: return call.noNecessaryDataError("remoteConfigurationId")

        qonversionSandwich.attachUserToRemoteConfiguration(remoteConfigurationId, call.toResultListener())
    }

    @PluginMethod
    fun detachUserFromRemoteConfiguration(call: PluginCall) {
        val remoteConfigurationId = call.getString("remoteConfigurationId") ?: return call.noNecessaryDataError("remoteConfigurationId")

        qonversionSandwich.detachUserFromRemoteConfiguration(remoteConfigurationId, call.toResultListener())
    }

    @PluginMethod
    fun userInfo(call: PluginCall) {
        qonversionSandwich.userInfo(call.toResultListener())
    }

    @PluginMethod
    fun storeSdkInfo(call: PluginCall) {
        val version = call.getString("version") ?: return call.noNecessaryDataError("version")
        val source = call.getString("source") ?: return call.noNecessaryDataError("source")

        qonversionSandwich.storeSdkInfo(source, version)
        call.resolve()
    }
}

File: ios/Sources/QonversionPlugin/QonversionPlugin.swift
import Foundation
import Capacitor
import QonversionSandwich

@objc(QonversionPlugin)
public class QonversionPlugin: CAPPlugin, CAPBridgedPlugin {
  public let identifier = "QonversionPlugin"
  public let jsName = "Qonversion"
  public let pluginMethods: [CAPPluginMethod] = [
    CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "identify", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "products", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "purchase", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "promoPurchase", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "checkEntitlements", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "remoteConfig", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "remoteConfigList", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "userInfo", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "userProperties", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "restore", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "offerings", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "setDefinedUserProperty", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "setCustomUserProperty", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "checkTrialIntroEligibility", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "attachUserToExperiment", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "detachUserFromExperiment", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "attachUserToRemoteConfiguration", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "detachUserFromRemoteConfiguration", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "storeSdkInfo", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "addAttributionData", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "collectAppleSearchAdsAttribution", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "isFallbackFileAccessible", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "presentCodeRedemptionSheet", returnType: CAPPluginReturnPromise)
  ]
  
  var qonversionSandwich: QonversionSandwich?
  
  override public func load() {
    qonversionSandwich = QonversionSandwich.init(qonversionEventListener: self)
  }
  
  @objc func initialize(_ call: CAPPluginCall) {
    guard let projectKey = call.getString("projectKey"),
          let launchMode = call.getString("launchMode") else {
      return call.noNecessaryDataError()
    }
    let environment = call.getString("environment")
    let entitlementsCacheLifetime = call.getString("entitlementsCacheLifetime")
    let proxyUrl = call.getString("proxyUrl")
    
    qonversionSandwich?.initialize(
      projectKey: projectKey,
      launchModeKey: launchMode,
      environmentKey: environment,
      entitlementsCacheLifetimeKey: entitlementsCacheLifetime,
      proxyUrl: proxyUrl
    )
    
    call.resolve()
  }
  
  @objc func identify(_ call: CAPPluginCall) {
    guard let userId = call.getString("userId") else {
      return call.noNecessaryDataError()
    }
    
    qonversionSandwich?.identify(userId, getDefaultCompletion(call))
  }
  
  @objc func products(_ call: CAPPluginCall) {
    qonversionSandwich?.products(getDefaultCompletion(call))
  }
  
  @objc func purchase(_ call: CAPPluginCall) {
    guard let productId = call.getString("productId") else {
      return call.noNecessaryDataError()
    }
    let quantity = call.getInt("quantity") ?? 1
      let contextKeys = call.getArray("contextKeys")?.capacitor.replacingNullValues().compactMap({$0}) as? [String] ?? []
    
    qonversionSandwich?.purchase(productId, quantity:quantity, contextKeys:contextKeys, completion: getDefaultCompletion(call))
  }
  
  @objc func promoPurchase(_ productId: String?, _ call: CAPPluginCall) {
    guard let productId = call.getString("productId") else {
      return call.noNecessaryDataError()
    }
    
    qonversionSandwich?.promoPurchase(productId, completion: getDefaultCompletion(call))
  }
  
  @objc func checkEntitlements(_ call: CAPPluginCall) {
    qonversionSandwich?.checkEntitlements(getDefaultCompletion(call))
  }
  
  @objc func remoteConfig(_ contextKey: String?, _ call: CAPPluginCall) {
    qonversionSandwich?.remoteConfig(contextKey, getDefaultCompletion(call))
  }
  
  @objc func remoteConfigList(_ call: CAPPluginCall) {
      guard let contextKeys = call.getArray("contextKeys")?.capacitor.replacingNullValues().compactMap({$0}) as? [String] else {
      qonversionSandwich?.remoteConfigList(getDefaultCompletion(call))
        return
    }
    
    guard let includeEmptyContextKey = call.getBool("includeEmptyContextKey") else {
      return call.noNecessaryDataError()
    }
    
    qonversionSandwich?.remoteConfigList(contextKeys, includeEmptyContextKey:includeEmptyContextKey, getDefaultCompletion(call))
  }
  
  @objc func userInfo(_ call: CAPPluginCall) {
    qonversionSandwich?.userInfo(getDefaultCompletion(call))
  }
  
  @objc func userProperties(_ call: CAPPluginCall) {
    qonversionSandwich?.userProperties(getDefaultCompletion(call))
  }
  
  @objc func restore(_ call: CAPPluginCall) {
    qonversionSandwich?.restore(getDefaultCompletion(call))
  }
  
  @objc func offerings(_ call: CAPPluginCall) {
    qonversionSandwich?.offerings(getDefaultCompletion(call))
  }
  
  @objc func setDefinedUserProperty(_ call: CAPPluginCall) {
    guard let rawProperty = call.getString("property"),
          let value = call.getString("value") else {
      return call.noNecessaryDataError()
    }
    
    qonversionSandwich?.setDefinedProperty(rawProperty, value: value)
    call.resolve()
  }
  
  @objc func setCustomUserProperty(_ call: CAPPluginCall) {
    guard let property = call.getString("property"),
          let value = call.getString("value") else {
      return call.noNecessaryDataError()
    }
    
    qonversionSandwich?.setCustomProperty(property, value: value)
    call.resolve()
  }
  
  @objc func checkTrialIntroEligibility(_ call: CAPPluginCall) {
      guard let ids = call.getArray("ids")?.capacitor.replacingNullValues().compactMap({$0}) as? [String] else {
      return call.noNecessaryDataError()
    }
    
    qonversionSandwich?.checkTrialIntroEligibility(ids, completion: getDefaultCompletion(call))
  }
  
  @objc func attachUserToExperiment(_ call: CAPPluginCall) {
    guard let experimentId = call.getString("experimentId"),
          let groupId = call.getString("groupId") else {
      return call.noNecessaryDataError()
    }
    
    qonversionSandwich?.attachUserToExperiment(with: experimentId, groupId: groupId, completion: getDefaultCompletion(call))
  }
  
  @objc func detachUserFromExperiment(_ call: CAPPluginCall) {
    guard let experimentId = call.getString("experimentId") else {
      return call.noNecessaryDataError()
    }
    
    qonversionSandwich?.detachUserFromExperiment(with: experimentId, completion: getDefaultCompletion(call))
  }
  
  @objc func attachUserToRemoteConfiguration(_ call: CAPPluginCall) {
    guard let remoteConfigurationId = call.getString("remoteConfigurationId") else {
      return call.noNecessaryDataError()
    }
    
    qonversionSandwich?.attachUserToRemoteConfiguration(with: remoteConfigurationId, completion: getDefaultCompletion(call))
  }
  
  @objc func detachUserFromRemoteConfiguration(_ call: CAPPluginCall) {
    guard let remoteConfigurationId = call.getString("remoteConfigurationId") else {
      return call.noNecessaryDataError()
    }
    
    qonversionSandwich?.detachUserFromRemoteConfiguration(with: remoteConfigurationId, completion: getDefaultCompletion(call))
  }
  
  @objc func storeSdkInfo(_ call: CAPPluginCall) {
    guard let version = call.getString("version"),
          let source = call.getString("source") else {
      return call.noNecessaryDataError()
    }
    
    qonversionSandwich?.storeSdkInfo(source: source, version: version)
    call.resolve()
  }
  
  @objc func addAttributionData(_ call: CAPPluginCall) {
      guard let data = call.getObject("data"),
          let provider = call.getString("provider") else {
      return call.noNecessaryDataError()
    }
    
    qonversionSandwich?.attribution(providerKey: provider, value: data)
    call.resolve()
  }
  
  @objc func collectAppleSearchAdsAttribution(_ call: CAPPluginCall) {
    qonversionSandwich?.collectAppleSearchAdsAttribution()
    call.resolve()
  }
  
  @objc func isFallbackFileAccessible(_ call: CAPPluginCall) {
    qonversionSandwich?.isFallbackFileAccessible(completion: getDefaultCompletion(call))
  }
  
  @objc func presentCodeRedemptionSheet(_ call: CAPPluginCall) {
    if #available(iOS 14.0, *) {
      qonversionSandwich?.presentCodeRedemptionSheet()
    }
    call.resolve()
  }
  
  private func getDefaultCompletion(_ call: CAPPluginCall) -> BridgeCompletion {
    return { data, error in
      if let error = error {
        return call.sandwichError(error)
      }
      
        guard let data else { return call.resolve() }
        
        call.resolve(data)
    }
  }
}

extension QonversionPlugin: QonversionEventListener {
  public func qonversionDidReceiveUpdatedEntitlements(_ entitlements: [String : Any]) {
    self.notifyListeners("entitlementsUpdatedEvent", data: entitlements)
  }
  
  public func shouldPurchasePromoProduct(with productId: String) {
    self.notifyListeners("shouldPurchasePromoProductEvent", data: ["productId": productId])
  }
}
